src/pkg/cli/compose/loader.go

```diff
@@ -86,6 +90,19 @@ func (c Loader) LoadProject(ctx context.Context) (*Project, error) {
 		return nil, err
 	}
 
+	// set interpolation config for interpolation
+	interpolationConfig.Environment = project.Environment
+
+	// load second time with the interpolation config now populated with env variables
+	project, err = projOpts.LoadProject(ctx)
+	if err != nil {
+		if errors.Is(err, errdefs.ErrNotFound) {
+			return nil, types.ErrComposeFileNotFound
+		}
+
+		return nil, err
+	}
```

Comment:
So we are loading twice, the first to get the environment mapping. The environment mapping is used in the custom substitute function.